--- a/linux/include/openswan/ipsec_esp.h
+++ b/linux/include/openswan/ipsec_esp.h
@@ -65,6 +65,11 @@ struct esphdr
 	__u8	esp_iv[8];		/* iv */
 };
 
+struct ipsec_skb_cb {
+	struct net_device *tunnel_dev;
+	struct net_device *eth_dev;
+};
+
 extern struct xform_functions esp_xform_funcs[];
 
 extern enum ipsec_rcv_value ipsec_rcv_esp_post_decrypt(struct ipsec_rcv_state *irs);
--- a/linux/net/ipsec/ipsec_esp.c
+++ b/linux/net/ipsec/ipsec_esp.c
@@ -165,6 +165,10 @@ ipsec_rcv_esp_authcalc(struct ipsec_rcv_
 		SHA1_CTX	sha1;
 	} tctx;
 
+	struct ipsecpriv *prv = netdev_to_ipsecpriv(skb->dev);
+	((struct ipsec_skb_cb *)skb->cb)->tunnel_dev = skb->dev;
+	((struct ipsec_skb_cb *)skb->cb)->eth_dev = prv->dev;
+
 #ifdef CONFIG_KLIPS_OCF
 	if (irs->ipsp->ocf_in_use)
 		return(ipsec_ocf_rcv(irs));
--- a/linux/net/ipsec/ipsec_tunnel.c
+++ b/linux/net/ipsec/ipsec_tunnel.c
@@ -2146,7 +2146,7 @@ ipsec_tunnel_init(struct net_device *dev
 	dev->hard_header_len 	= 0;
 	dev->mtu		= 0;
 	dev->addr_len		= 0;
-	dev->type		= ARPHRD_VOID; /* ARPHRD_TUNNEL; */ /* ARPHRD_ETHER; */
+	dev->type		= 31; /* ARPHRD_IPSEC_TUNNEL */
 	dev->tx_queue_len	= 10;		/* Small queue */
 #ifdef IFF_XMIT_DST_RELEASE
 	dev->priv_flags	       &= ~IFF_XMIT_DST_RELEASE;
@@ -2590,7 +2590,7 @@ ipsec_tunnel_attach(struct net_device *d
 	prv->mtu = physdev->mtu;
 
 #ifdef PHYSDEV_TYPE
-	dev->type = physdev->type; /* ARPHRD_TUNNEL; */
+	dev->type = 31;
 #endif /*  PHYSDEV_TYPE */
 
 	dev->addr_len = physdev->addr_len;
--- a/linux/net/ipsec/ipsec_xmit.c
+++ b/linux/net/ipsec/ipsec_xmit.c
@@ -940,7 +940,10 @@ ipsec_xmit_esp(struct ipsec_xmit_state *
 		osw_ip6_hdr(ixs)->nexthdr = IPPROTO_ESP;
 	else
 		osw_ip4_hdr(ixs)->protocol = IPPROTO_ESP;
-	
+
+	((struct ipsec_skb_cb *)ixs->skb->cb)->tunnel_dev = ixs->dev;
+	((struct ipsec_skb_cb *)ixs->skb->cb)->eth_dev = ixs->iprv->dev;
+
 #ifdef CONFIG_KLIPS_OCF
 	if (ixs->ipsp->ocf_in_use) {
 		/* handle the IV code here for now,  near the similar code below */
