Index: uboot-1.0-2012.07/common/env_common.c
===================================================================
--- uboot-1.0-2012.07.orig/common/env_common.c
+++ uboot-1.0-2012.07/common/env_common.c
@@ -214,6 +214,8 @@ int env_import(const char *buf, int chec
 
 		if (crc32(0, ep->data, ENV_SIZE) != crc) {
 			set_default_env("!bad CRC");
+			puts("Saving default environment into flash\n\n");
+			saveenv();
 			return 0;
 		}
 	}
